### PR TITLE
JAMES-1716 Saving to draft should not send message

### PR DESCRIPTION
**Pull request by fvn-linagora**
When saving a message to drafts mailbox, server should not send it to the specified recipients.

SetMessageCreationProcessor is getting mailboxes through an adapter (SystemMailboxesProvider) which makes stubbing for unit testing easier than when coupled with MailboxManager.

providing both an outside integration test (SetMessageMethodTest) and some unit tests (SetMessageCreationProcessorTest)

Note: previous feedbacks (from https://github.com/linagora/james-project/pull/203) have been applied on this PR

### CODE REVIEW

File: server/container/guice/memory-guice/src/main/java/org/apache/james/modules/mailbox/MemoryMailboxModule.java [hunk 1]
```diff
@@ -78,6 +80,7 @@ protected void configure() {
         bind(MailboxManager.class).to(InMemoryMailboxManager.class);
         bind(MailboxACLResolver.class).to(UnionMailboxACLResolver.class);
         bind(GroupMembershipResolver.class).to(SimpleGroupMembershipResolver.class);
+        bind(new TypeLiteral<SystemMailboxesProvider<InMemoryId>>(){}).to(new TypeLiteral<SystemMailboxesProviderImpl<InMemoryId>>(){});
```
Comment (chibenwa) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/mailbox/MemoryMailboxModule.java]: It's JMAP stuff. Shouldn't it be located in a JMAP module ?

**Comment by chibenwa** [comment]
:+1:  for me, except for the role resolver, which should be injected from within a JMAP module.

File: server/container/guice/memory-guice/src/main/java/org/apache/james/modules/mailbox/MemoryMailboxModule.java (hunk 1 above)
Comment (fvn-linagora) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/mailbox/MemoryMailboxModule.java]: Ideally yes.. But as there is no Cassandra/InMemory guice module in JMAP, I have to put registration in here I guess.

Comment (chibenwa) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/mailbox/MemoryMailboxModule.java]: I don't agree with you.

-> It's the only includes in here that are not in Mailbox

We had the same kind of problems with Methods during memory / cassandra project split.

Matthieu proposed this :

```
methods.addBinding().to(guiceGenericType.newGenericType(GetMailboxesMethod.class));
```

Seems nicer, in my opinion.

Comment (fvn-linagora) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/mailbox/MemoryMailboxModule.java]: OK, I finally managed to put bindings for generic types in a more appropriate guice module (JMAPModule), thanks to Antoine.
Note that I personally wouldn't have used the word "nice" in such a context, but I do understand your point :smiley:

**Comment by aduprat** [comment]
:+1:

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 2]
```diff
@@ -1345,6 +1346,60 @@ public void setMessagesShouldSendAReadableHtmlMessage() throws Exception {
...
+                "  ]" +
+                "]";
+
+        // Given
+        given()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+                .header("Authorization", this.accessToken.serialize())
+                .body(requestBody)
+                // When
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Ident

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java [hunk 3]
```diff
@@ -126,12 +122,58 @@ public SetMessagesResponse process(SetMessagesRequest request, MailboxSession ma
...
+                .collect(Collectors.toMap(Map.Entry::getKey, e -> notImplementedError));
+    }
+
+    private List<AbstractMap.SimpleEntry<CreationMessageId, SetError>> getInvalidArgumentsErrors(
+            Set<Map.Entry<CreationMessageId, CreationMessage>> creationRequestEntries,
+            Predicate<CreationMessage> invalidMessagesTester, Map<CreationMessageId, SetError> mapOfNotImplEntries) {
+
+        // Deal with invalid arguments but only for entries that have not been flagged yet (ie not impl...)
+        return creationRequestEntries.stream()
+                .filter(e -> !mapOfNotImplEntries.containsKey(e.getKey())) // exclude requests that have already been flagged with "not implemented"
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java]: I really don't like this pattern, but ok for this time, should refactor next time.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SystemMailboxesProvider.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.methods;
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SystemMailboxesProvider.java]: That's not a method.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SystemMailboxesProviderImpl.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.methods;
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SystemMailboxesProviderImpl.java]: Neither.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetError.java [hunk 6]
```diff
@@ -102,18 +102,26 @@ public String getType() {
...
     @Override
-    public boolean equals(Object obj) {
-        if (obj instanceof SetError) {
-            SetError other = (SetError) obj;
-            return Objects.equal(this.type, other.type)
-                && Objects.equal(this.description, other.description)
-                && Objects.equal(this.properties, other.properties);
-        }
-        return false;
+    public boolean equals(Object o) {
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetError.java]: Why this change?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesCreationProcessorTest.java [hunk 7]
```diff
@@ -129,69 +144,42 @@ public void processShouldReturnNonEmptyCreatedWhenRequestHasNonEmptyCreate() thr
...
-            }
-        };
+        TestSystemMailboxesProvider stubbedMailboxes = TestSystemMailboxesProvider.fakeOutbox(() -> Optional.of(fakeOutbox));
+        SetMessagesCreationProcessor<TestId> sut = new SetMessagesCreationProcessor<>(
+                stubSessionMapperFactory, new MIMEMessageConverter(), mockedMailSpool, mockedMailFactory, stubbedMailboxes);
         // When
         sut.process(buildFakeCreationRequest(), buildStubbedSession());
 
         // Then
+        //noinspection unchecked
```
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesCreationProcessorTest.java]: ??

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesCreationProcessorTest.java [hunk 8]
```diff
@@ -205,9 +193,134 @@ private SetMessagesRequest buildFakeCreationRequest() {
...
+    @Test
+    public void processShouldNotSpoolMailWhenNotSavingToOutbox() throws Exception {
+        // Given
+        SetMessagesCreationProcessor<TestId> sut = new SetMessagesCreationProcessor<>(
+                stubSessionMapperFactory, new MIMEMessageConverter(), mockedMailSpool, mockedMailFactory, fakeSystemMailboxesProvider);
+        // When
+        sut.process(buildCreationRequestNotForSending(), buildStubbedSession());
+
+        // Then
+        verify(mockedMailSpool, never()).send(any(Mail.class), any(MailMetadata.class));
```
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesCreationProcessorTest.java]: How do you see when reading this test that you are not saving to outbox?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 2 above)
Comment (fvn-linagora) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: fixed

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetError.java (hunk 6 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetError.java]: a merge/rebase conflict of 2 concurrent branches that require SetError comparison (my branch has been stuck in the review pipeline for longer!). equals() method had no real reason to be replaced (except for resolving a merge conflict).
toString was used for unit testing purpose IIRC.

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesCreationProcessorTest.java (hunk 7 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesCreationProcessorTest.java]: this is a way to suppress reported compilation/style warnings for a given statement :
add's 2nd parameter is expecting a generic but is provided with a non generic one (any(MailboxMessage.class))
I can remove the line

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SystemMailboxesProvider.java (hunk 4 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SystemMailboxesProvider.java]: Locality is important and there is only one use by a single method. I can move it to utils namespace if that looks like a better place.

**Comment by rouazana** [comment]
:+1:

**Comment by chibenwa** [comment]
:+1:

**Comment by aduprat** [comment]
:+1:

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProvider.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+import java.util.stream.Stream;
+
+import org.apache.james.jmap.model.mailbox.Role;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.store.mail.model.Mailbox;
+import org.apache.james.mailbox.store.mail.model.MailboxId;
+
+public interface SystemMailboxesProvider<Id extends MailboxId> {
+    Stream<Mailbox<Id>> getStreamOfMailboxesFromRole(Role aRole, MailboxSession session);
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProvider.java]: listMailboxes(MailboxSession session, Role role)

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProviderImpl.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    private boolean hasRole(Role aRole, MailboxPath mailBoxPath) {
+        return Role.from(mailBoxPath.getName())
+                .map(aRole::equals)
+                .orElse(false);
+    }
+
+    public Stream<Mailbox<Id>> getStreamOfMailboxesFromRole(Role aRole, MailboxSession session) {
+        ThrowingSupplier<List<MailboxMetaData>> getAllMailboxes = () -> mailboxManager.search(MailboxQuery.builder(session).privateUserMailboxes().build(), session);
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProviderImpl.java]: why do you need this supplier ?

**Comment by fvn-linagora** [comment]
test this please

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProvider.java (hunk 9 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProvider.java]: renamed

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProviderImpl.java (hunk 10 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProviderImpl.java]: to avoid littering the call tree with useless checked exception handling

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java [hunk 11]
```diff
@@ -78,60 +75,106 @@
...
+                                        Predicate<CreationMessage> validMessagesTester) {
+
+        boolean isMessageSetInOutbox = systemMailboxesProvider.getStreamOfMailboxesFromRole(Role.OUTBOX, session).findFirst()
+                .map(box -> box.getMailboxId().serialize())
+                .map(id -> messageWithId.getMailboxIds().contains(id))
+                .orElse(false);
+        return validMessagesTester.test(messageWithId) && isMessageSetInOutbox;
+    }
+
+    private Predicate<CreationMessage> getNotImplementedRequestTester(MailboxSession mailboxSession) {
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java]: Why getNotImplementedRequestTester ?

**Comment by chibenwa** [comment]
:+1:

**Comment by chibenwa** [comment]
Duplicate of https://github.com/linagora/james-project/pull/208 ?

**Comment by mbaechler** [comment]
done in another PR.
